src/modules/cmdpal/ext/Microsoft.CmdPal.Ext.Apps/AllAppsPage.cs

```diff
@@ -37,89 +47,155 @@ public AllAppsPage()
 
-        stopwatch.Stop();
-        Logger.LogTrace($"{nameof(AllAppsPage)}.{nameof(BuildListItems)} took: {stopwatch.ElapsedMilliseconds} ms");
+                stopwatch.Stop();
+                Logger.LogTrace($"{nameof(AllAppsPage)}.{nameof(BuildListItems)} took: {stopwatch.ElapsedMilliseconds} ms");
+            }
+        }
     }
 
-    internal List<AppItem> GetPrograms()
+    internal (AppItem[] AllApps, AppListItem[] PinnedItems, AppListItem[] UnpinnedItems) GetPrograms()
     {
         var uwpResults = AppCache.Instance.Value.UWPs
             .Where((application) => application.Enabled)
-            .Select(UwpToAppItem);
+            .Select(app => app.ToAppItem());
 
         var win32Results = AppCache.Instance.Value.Win32s
             .Where((application) => application.Enabled && application.Valid)
-            .Select(app =>
+            .Select(app => app.ToAppItem());
+
+        var allApps = uwpResults.Concat(win32Results).ToList();
+
+        var pinned = new List<AppListItem>();
+        var unpinned = new List<AppListItem>();
+
+        foreach (var app in allApps)
+        {
+            var isPinned = PinnedAppsManager.Instance.IsAppPinned(app.AppIdentifier);
+            app.Commands?.AddRange(AddPinCommands(app, isPinned));
+
+            var appListItem = new AppListItem(app, true);
+
+            if (isPinned)
             {
-                var icoPath = string.IsNullOrEmpty(app.IcoPath) ?
-                    (app.AppType == Win32Program.ApplicationType.InternetShortcutApplication ?
-                        app.IcoPath :
-                        app.FullPath) :
-                    app.IcoPath;
-
-                // icoPath = icoPath.EndsWith(".lnk", System.StringComparison.InvariantCultureIgnoreCase) ? (icoPath + ",0") : icoPath;
-                icoPath = icoPath.EndsWith(".lnk", System.StringComparison.InvariantCultureIgnoreCase) ?
-                    app.FullPath :
-                    icoPath;
-                return new AppItem()
-                {
-                    Name = app.Name,
-                    Subtitle = app.Description,
-                    Type = app.Type(),
-                    IcoPath = icoPath,
-                    ExePath = !string.IsNullOrEmpty(app.LnkFilePath) ? app.LnkFilePath : app.FullPath,
-                    DirPath = app.Location,
-                    Commands = app.GetCommands(),
-                };
-            });
-
-        return uwpResults.Concat(win32Results).OrderBy(app => app.Name).ToList();
+                appListItem.Tags = appListItem.Tags
+                                            .Concat([new Tag() { Icon = Icons.PinIcon }])
+                                            .ToArray();
+                pinned.Add(appListItem);
+            }
+            else
+            {
+                unpinned.Add(appListItem);
+            }
+        }
+
+        return (
+                allApps
+                    .ToArray(),
```

Comment:
technically we don't use the `.AllApps` member of this tuple anywhere, do we? We could probably save this one `.ToArray()` call by dropping it, right?

Reply:
Oh so glad you caught that. I was wondering why it seemed to respond slowly to pin/unpin. It's because I wasn't using that array so it was actually pulling all the apps from the cache and building the list of appitems and applistitems over again.